parser-typechecker/src/Unison/Typechecker/Components.hs

```diff
@@ -35,25 +34,26 @@ ordered = ABT.orderedComponents
 --
 -- Fails on the left if there are duplicate definitions.
 minimize ::
-  (Var v) =>
+  forall vt v a.
+  (Var v, Ord a) =>
   Term' vt v a ->
-  Either (NonEmpty (v, [a])) (Maybe (Term' vt v a))
+  Either (NonEmpty (v, NESet a)) (Maybe (Term' vt v a))
 minimize (Term.LetRecNamedAnnotatedTop' isTop blockAnn bs e) =
-  let bindings = first snd <$> bs
-      group =
-        map (fst . head &&& map (ABT.annotation . snd))
-          . groupBy ((==) `on` fst)
-          . sortBy
-            (compare `on` fst)
-      grouped = group bindings
-      dupes = filter ok grouped
+  let bindings :: [(v, Term' vt v a)]
+      bindings = bs <&> \((_a, v), t) -> (v, t)
+      grouped :: Map v (NESet a)
+      grouped =
+        bs
+          & fmap (\((a, v), _t) -> (v, NESet.singleton a)) -- For duplicates, we care about the binding location.
+          & Map.fromListWith (NESet.union)
+      dupes = Map.filterWithKey ok grouped
         where
-          ok (v, as)
+          ok v as
             | Var.name v == "_" = False
             | otherwise = length as > 1
-   in if not $ null dupes
-        then Left $ Nel.fromList dupes
-        else
+   in case NEL.nonEmpty $ Map.toList dupes of
+        Just dupeList -> Left dupeList
+        Nothing -> do
```

Comment:
At one point I thought minimization was affecting how lets were being constructed; it turns out to be unrelated but I had a tough time reading this function as it was, so made a couple changes to help readability as well as to use stronger types like NESet instead of `[a]`